include/runtime/collections/rangemap.h

```diff
@@ -0,0 +1,426 @@
+  }
+
+  /* Given a range to be removed and this map, return a map that does not
+     * contain the removed range or any part of it.
+     * If the range to be removed partially overlaps with any ranges in the
+     * map, then the boundaries of these ranges are adjusted in the resulting
+     * map so that they do not overlap with the removed range.
+     */
+  RangeMap deleted(Range<T> const &r) const {
+    // Empty ranges do not make sense here.
+    if (r.is_empty()) {
+      CONSTRUCT_MSG_AND_THROW("Delete empty range from range map");
+    }
+
+    std::vector<std::pair<Range<T>, V>> ranges;
+    get_overlapping_ranges(treemap_, r, ranges);
+    // Each relevant range may lead to changes to the existing underlying
+    // treemap data structure.
+    // We iterate over the collected relevant ranges to collect and apply
+    // these changes.
+    T ds = r.get_start();
+    T de = r.get_end();
+    RBTree<Range<T>, V> tmpmap = treemap_;
+    for (auto &p : ranges) {
+      Range<T> rr = p.first;
+      V rv = p.second;
+      assert(r.overlaps(rr));
+      T rrs = rr.get_start();
+      T rre = rr.get_end();
+      tmpmap = tmpmap.deleted(rr);
+      if (rrs < ds) {
+        tmpmap = tmpmap.inserted(Range<T>(rrs, ds), rv);
+      }
+      if (rre > de) {
+        tmpmap = tmpmap.inserted(Range<T>(de, rre), rv);
+      }
+    }
+    return RangeMap(tmpmap);
+  }
+
+  // Return true if range r partially or completely overlaps with any key range
+  // stored in this rangemap.
+  bool overlaps(Range<T> const &r) const { return overlaps(treemap_, r); }
+
+  // Print this rangemap to output stream os.
+  void print(std::ostream &os) const {
+    os << "--------------------------" << std::endl;
+    for_each(treemap_, [&os](Range<T> x, V v) {
+      os << "[ " << x.get_start() << ".." << x.get_end() << " ) -> " << v
+         << std::endl;
+    });
+    os << "--------------------------" << std::endl;
+  }
+
+  // Return a rangemap that is the concatenation of this rangemap and rangemap
+  // m. Throw an exception if any of the key ranges in this rangemap overlaps
+  // with any of the key ranges in rangemap m.
+  RangeMap concat(RangeMap const &m) const {
+    RangeMap res = *this;
+    for_each(m.treemap_, [&res, this](Range<T> const &x, V const &v) {
+      if (!overlaps(x)) {
+        res = res.inserted(x, v);
+      } else {
+        CONSTRUCT_MSG_AND_THROW("Overlapping key ranges in map concatenation");
+      }
+    });
+    return res;
+  }
+};
+
+// Return a rangemap with all elements in m, and then also from the container
+// designated by the beginning and end iterator arguments. The container should
+// contain elements of type std::pair<Range<T>, V>.
+template <class T, class V, class I>
+RangeMap<T, V> inserted(RangeMap<T, V> const &m, I it, I end) {
+  if (it == end) {
+    return m;
+  }
+  Range<T> key = it->first;
+  V val = it->second;
+  auto m1 = m.inserted(key, val);
+  return inserted(m1, ++it, end);
+}
```

Comment:
See previous comment re. namespaces

Reply:
As above, the structure is enclosed in namespace `rangemap`.